util/blooms-db/src/lib.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Ethereum blooms database
+
+extern crate byteorder;
+extern crate ethbloom;
+extern crate parking_lot;
+extern crate tiny_keccak;
+
+#[cfg(test)]
+extern crate tempdir;
+
+mod db;
+mod file;
+
+use std::io;
+use std::path::Path;
+use parking_lot::Mutex;
+
+/// Threadsafe API for blooms database.
+///
+/// # Warning
+///
+/// This database does not guarantee atomic writes.
+pub struct Database {
+	database: Mutex<db::Database>,
```

Comment:
Any reasons we use `Mutex` but not `RwLock`? I think it would be safe for multiple parties to read blooms at the same time?

Reply:
Great question! There is a reason, https://users.rust-lang.org/t/how-to-handle-match-with-irrelevant-ok--/6291/15

Reply:
tl;dr

> Reading a file is mutating the object that represent the file stream, namely altering the position. The very data being mutated is stored in the kernel space, and simultaneous mutation won’t cause the kernel any trouble because it is smart enough to synchronize accesses.
Still, two threads simultaneously reading the same file stream will break internal logic. (I guess even a single read() call on BufReader could yield a non-contiguous slice).

Reply:
That's why I made iterator function of `db::Database` `mut`.

Reply:
Ah got it!